pkg/provisioner/fss/fss.go

```diff
@@ -84,175 +95,373 @@ func (fsp *filesystemProvisioner) getMountTargetFromID(ctx context.Context, moun
+	if target.ExportSetId == nil {
+		logger.Error("Mount target has no export set associated with it")
+		return nil, errors.Errorf("mount target has no export set associated with it")
+	}
+
+	// Randomnly select IP address associated with the mount target to use
+	// for attachment.
 	var serverIP string
 	{
-		ctx, cancel := context.WithTimeout(ctx, fsp.client.Timeout())
-		defer cancel()
 		id := target.PrivateIpIds[rand.Int()%len(target.PrivateIpIds)]
-		getPrivateIPResponse, err := fsp.client.VCN().GetPrivateIp(ctx, core.GetPrivateIpRequest{
-			PrivateIpId: &id,
-		})
+		getPrivateIPResponse, err := fsp.client.VCN().GetPrivateIp(ctx, core.GetPrivateIpRequest{PrivateIpId: &id})
 		if err != nil {
-			logger.With(zap.Error(err), "privateIPID", id).Errorf("Failed to retrieve IP address for mount target privateIpID=%q: %v", id, err)
+			logger.With(zap.Error(err), "privateIPID", id).Error("Failed to retrieve IP address for mount target")
 			return nil, err
 		}
 		serverIP = *getPrivateIPResponse.PrivateIp.IpAddress
 	}
 
-	region, ok := os.LookupEnv("OCI_SHORT_REGION")
-	if !ok {
-		metadata, err := fsp.metadata.Get()
+	logger.With("privateIP", serverIP).Info("Selected Mount Target Private IP.")
+
+	logger.Info("Creating FileSystem")
+	var fsID string
+	{
+		fs, err := fsp.getOrCreateFileSystem(ctx, logger, *ad.Name, fsDisplayName)
 		if err != nil {
 			return nil, err
 		}
-		region = metadata.Region
+		fsID = *fs.Id
+
+		logger = logger.With("fileSystemID", fsID)
 	}
 
-	logger.With("privateIP", serverIP).Info("Selected Mount Target Private IP.")
+	logger.Info("Creating export set")
+	export, err := fsp.getOrCreateExport(ctx, logger, fsID, *target.ExportSetId)
+	if err != nil {
+		logger.With(zap.Error(err)).Error("Failed to create export.")
+		return nil, err
+	}
 
 	return &v1.PersistentVolume{
 		ObjectMeta: metav1.ObjectMeta{
 			Name: fsID,
 			Annotations: map[string]string{
 				ociVolumeID: fsID,
-				ociExportID: exportSetID,
-			},
-			Labels: map[string]string{
-				plugin.LabelZoneRegion: region,
+				ociExportID: *export.Id,
 			},
+			Labels: map[string]string{plugin.LabelZoneRegion: region},
 		},
 		Spec: v1.PersistentVolumeSpec{
 			PersistentVolumeReclaimPolicy: options.PersistentVolumeReclaimPolicy,
 			AccessModes:                   options.PVC.Spec.AccessModes,
-			//FIXME: fs storage doesn't enforce quota, capacity is meaningless here.
+			//NOTE: fs storage doesn't enforce quota, capacity is meaningless here.
 			Capacity: v1.ResourceList{
 				v1.ResourceName(v1.ResourceStorage): options.PVC.Spec.Resources.Requests[v1.ResourceName(v1.ResourceStorage)],
 			},
 			PersistentVolumeSource: v1.PersistentVolumeSource{
 				NFS: &v1.NFSVolumeSource{
-					// Randomnly select IP address associated with the mount target to use for attachment
 					Server:   serverIP,
-					Path:     "/" + fsID,
-					ReadOnly: false,
+					Path:     *export.Path,
+					ReadOnly: false, // TODO: Should this be based on the AccessModes?
```

Comment:
Yes could be one of ReadOnlyMany or ReadWriteMany. We'd need a translation function to do this AFAIK.